### ISSUE TITLE
Strategies for reducing Database size

### ISSUE THREAD

**Issue by maxwell** [diaspora/diaspora#6604]
Hi All,

Looking into some ways to optimize the jd.com dataset for current users.  Also, its just interesting to see what *doesnt** scale as a pod gets larger.

here is a snapshot of the table sizes from mysql table status.

It looks like the participations table is a huge chunk (~25%) of total database size.  Does it make sense to expire those records at a point?

```text
+---------------------------+--------+---------+------------+----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-------------------+----------+----------------+---------+
| Name                      | Engine | Version | Row_format | Rows     | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time | Check_time | Collation         | Checksum | Create_options | Comment |
+---------------------------+--------+---------+------------+----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-------------------+----------+----------------+---------+
| account_deletions         | InnoDB |      10 | Compact    |    66660 |             71 |     4734976 |               0 |            0 |   4194304 |          76685 | 2015-07-02 07:35:46 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| aspect_memberships        | InnoDB |      10 | Compact    |  1138177 |             60 |    68780032 |               0 |     91553792 |   5242880 |        5902153 | 2012-01-09 19:28:03 | NULL        | NULL       | utf8_bin          |     NULL |                |         |
| aspect_visibilities       | InnoDB |      10 | Compact    | 12326295 |             51 |   636485632 |               0 |   1069481984 |   6291456 |       18440715 | 2015-07-02 07:41:25 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| aspects                   | InnoDB |      10 | Compact    |  1245718 |             52 |    65634304 |               0 |     76709888 |   4194304 |        1322095 | 2015-07-02 07:41:50 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| blocks                    | InnoDB |      10 | Compact    |    38200 |             41 |     1589248 |               0 |            0 |   4194304 |          48369 | 2012-01-09 19:33:31 | NULL        | NULL       | latin1_swedish_ci |     NULL |                |         |
| chat_contacts             | InnoDB |      10 | Compact    |        0 |              0 |       16384 |               0 |        16384 |         0 |              1 | 2015-07-02 07:41:50 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| chat_fragments            | InnoDB |      10 | Compact    |        0 |              0 |       16384 |               0 |        16384 |         0 |              1 | 2015-07-02 07:41:50 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| chat_offline_messages     | InnoDB |      10 | Compact    |        0 |              0 |       16384 |               0 |            0 |         0 |              1 | 2015-07-02 07:41:50 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| comments                  | InnoDB |      10 | Compact    |  4152982 |           1354 |  5624561664 |               0 |    325058560 |   6291456 |        4130853 | 2015-07-02 07:51:05 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| contacts                  | InnoDB |      10 | Compact    |  1820644 |             58 |   106561536 |               0 |     92323840 |   5242880 |        2154317 | 2012-01-09 19:41:40 | NULL        | NULL       | utf8_bin          |     NULL |                |         |
| conversation_visibilities | InnoDB |      10 | Compact    |    88179 |             65 |     5783552 |               0 |      6766592 |   4194304 |         105591 | 2012-01-09 19:42:54 | NULL        | NULL       | utf8_bin          |     NULL |                |         |
| conversations             | InnoDB |      10 | Compact    |    45164 |             81 |     3686400 |               0 |      1589248 |   3145728 |          48540 | 2015-07-02 07:51:07 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| invitation_codes          | InnoDB |      10 | Compact    |   145382 |             68 |     9977856 |               0 |            0 |   4194304 |         146049 | 2015-07-02 07:51:10 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| invitations               | InnoDB |      10 | Compact    |   537619 |             88 |    47792128 |               0 |     34127872 |   5242880 |         627355 | 2015-07-02 07:51:24 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| likes                     | InnoDB |      10 | Compact    |  4941870 |           1733 |  8564768768 |               0 |    764772352 |   7340032 |        6504117 | 2015-07-02 08:07:10 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| locations                 | InnoDB |      10 | Compact    |     2363 |            173 |      409600 |               0 |            0 |         0 |           2590 | 2015-07-02 08:07:12 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| mentions                  | InnoDB |      10 | Compact    |    75044 |             46 |     3473408 |               0 |      7421952 |   4194304 |         110966 | 2012-01-09 19:49:03 | NULL        | NULL       | utf8_bin          |     NULL |                |         |
| messages                  | InnoDB |      10 | Compact    |   187493 |           2392 |   448643072 |               0 |     12615680 |   5242880 |         227002 | 2015-07-02 08:08:09 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| notification_actors       | InnoDB |      10 | Compact    |  5523558 |             57 |   316456960 |               0 |    438108160 |   6291456 |        7267269 | 2012-01-09 19:49:47 | NULL        | NULL       | utf8_bin          |     NULL |                |         |
| notifications             | InnoDB |      10 | Compact    |  3887505 |             81 |   317505536 |               0 |    326238208 |   5242880 |        5499753 | 2015-07-02 08:10:25 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| o_embed_caches            | InnoDB |      10 | Compact    |   278344 |            793 |   220938240 |               0 |     27918336 |   4194304 |         325353 | 2015-07-02 08:10:48 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| open_graph_caches         | InnoDB |      10 | Compact    |   547269 |            526 |   288096256 |               0 |            0 |   6291456 |         588947 | 2015-07-02 08:11:14 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| participations            | InnoDB |      10 | Compact    |  6637022 |           1784 | 11842617344 |               0 |    690962432 |   5242880 |        9353579 | 2015-07-02 08:36:30 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| people                    | InnoDB |      10 | Compact    |   483795 |           1074 |   520060928 |               0 |     66797568 |   6291456 |         618200 | 2015-07-02 08:37:49 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| photos                    | InnoDB |      10 | Compact    |  1417487 |            226 |   321732608 |               0 |     80412672 |   4194304 |        1780954 | 2015-07-02 08:38:52 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| pods                      | InnoDB |      10 | Compact    |     1889 |             78 |      147456 |               0 |            0 |         0 |           2066 | 2015-07-02 08:38:53 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| poll_answers              | InnoDB |      10 | Compact    |     5558 |             94 |      524288 |               0 |        98304 |   4194304 |           6395 | 2015-07-02 08:38:54 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| poll_participations       | InnoDB |      10 | Compact    |    10112 |           1920 |    19415040 |               0 |       294912 |   4194304 |          13671 | 2015-07-02 08:38:56 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| polls                     | InnoDB |      10 | Compact    |     1573 |            135 |      212992 |               0 |        49152 |         0 |           1773 | 2015-07-02 08:38:56 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| posts                     | InnoDB |      10 | Compact    |  3933019 |            703 |  2765094912 |               0 |   1092026368 |   7340032 |        6933763 | 2015-07-02 08:52:24 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| profiles                  | InnoDB |      10 | Compact    |   550342 |            319 |   175865856 |               0 |     46694400 |   4194304 |         572052 | 2015-07-02 08:53:11 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| rails_admin_histories     | InnoDB |      10 | Compact    |        0 |              0 |       16384 |               0 |        16384 |         0 |              1 | 2015-07-02 08:53:11 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| reports                   | InnoDB |      10 | Compact    |      129 |            127 |       16384 |               0 |        16384 |         0 |            149 | 2015-07-02 08:53:11 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| roles                     | InnoDB |      10 | Compact    |       53 |            309 |       16384 |               0 |            0 |         0 |             68 | 2015-07-02 08:53:11 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| schema_migrations         | InnoDB |      10 | Compact    |      162 |            101 |       16384 |               0 |        16384 |         0 |           NULL | 2015-07-02 08:53:11 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| services                  | InnoDB |      10 | Compact    |   184295 |            196 |    36241408 |               0 |     17907712 |   5242880 |         214182 | 2015-07-02 08:53:18 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| share_visibilities        | InnoDB |      10 | Compact    | 34905026 |             58 |  2027945984 |               0 |   4801363968 |   7340032 |      117779114 | 2015-07-02 09:19:17 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| simple_captcha_data       | InnoDB |      10 | Compact    |       85 |            192 |       16384 |               0 |        16384 |         0 |          42528 | 2015-07-02 09:19:18 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| tag_followings            | InnoDB |      10 | Compact    |   725115 |             55 |    40435712 |               0 |     56868864 |   6291456 |         856073 | 2012-01-10 01:22:35 | NULL        | NULL       | utf8_bin          |     NULL |                |         |
| taggings                  | InnoDB |      10 | Compact    |  8646215 |             51 |   442400768 |               0 |   1137950720 |   6291456 |       11228100 | 2015-07-02 09:24:34 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| tags                      | InnoDB |      10 | Compact    |   697012 |             44 |    30998528 |               0 |     26820608 |   4194304 |         763604 | 2015-07-02 09:24:48 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| user_preferences          | InnoDB |      10 | Compact    |   104370 |             65 |     6832128 |               0 |            0 |   4194304 |         121773 | 2015-07-02 09:24:51 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
| users                     | InnoDB |      10 | Compact    |   600736 |           5510 |  3310354432 |               0 |     77250560 |   7340032 |         805596 | 2015-07-02 10:09:26 | NULL        | NULL       | utf8mb4_bin       |     NULL |                |         |
+---------------------------+--------+---------+------------+----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-------------------+----------+----------------+---------+
43 rows in set (0.38 sec)
```

**Comment by maxwell** [comment]
Here is a breakdown of the 10 biggest tables

```text
+-----------------------------------------+--------+--------+-------+------------+---------+
| CONCAT(table_schema, '.', table_name)   | rows   | DATA   | idx   | total_size | idxfrac |
+-----------------------------------------+--------+--------+-------+------------+---------+
| diaspora_production.participations      | 6.64M  | 11.03G | 0.64G | 11.67G     |    0.06 |
| diaspora_production.likes               | 4.94M  | 7.98G  | 0.71G | 8.69G      |    0.09 |
| diaspora_production.share_visibilities  | 34.91M | 1.89G  | 4.47G | 6.36G      |    2.37 |
| diaspora_production.comments            | 4.15M  | 5.24G  | 0.30G | 5.54G      |    0.06 |
| diaspora_production.posts               | 3.93M  | 2.58G  | 1.02G | 3.59G      |    0.39 |
| diaspora_production.users               | 0.60M  | 3.08G  | 0.07G | 3.15G      |    0.02 |
| diaspora_production.aspect_visibilities | 12.33M | 0.59G  | 1.00G | 1.59G      |    1.68 |
| diaspora_production.taggings            | 8.65M  | 0.41G  | 1.06G | 1.47G      |    2.57 |
| diaspora_production.notification_actors | 5.52M  | 0.29G  | 0.41G | 0.70G      |    1.38 |
| diaspora_production.notifications       | 3.89M  | 0.30G  | 0.30G | 0.60G      |    1.03 |
+-----------------------------------------+--------+--------+-------+------------+---------+

```

**Comment by denschub** [comment]
Actually duplicate of #4920, closing but leaving a reference over there...

**Comment by jaywink** [comment]
Yeah @maxwell we've been pinging your thoughts re this a few times ;)

**Comment by Flaburgan** [comment]
Thank you for these data @maxwell! @cmrd-senya is working on a patch (https://github.com/diaspora/diaspora/pull/6586) to reduce the size of the `comments`, `poll_participations`, `messages`, `participations` and `likes` tables.

**Comment by jaywink** [comment]
Also there is the db cleaner PR to clean whatever is left ;) Should drop joindiaspora.com quite significantly.

**Comment by maxwell** [comment]
Jason,

Id love some help in administering said database cleaner if you had time.

- maxwell

On Wed, Dec 30, 2015 at 7:51 AM, Jason Robinson <notifications@github.com>
wrote:

> Also there is the db cleaner PR to clean whatever is left ;) Should drop
> joindiaspora.com quite significantly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diaspora/diaspora/issues/6604#issuecomment-168022807>.
>



-- 
maxwell salzberg - BackerKit, founder
skype: max.salzberg
p: (203) 451-3600
email: maxwell@backerkit.com
twitter: @maxwellsalz

**Comment by cmrd-senya** [comment]
I doubt we really need a DB cleaner after we remove all the spare signatures from the database. Starting with merging #6586.

**Comment by jaywink** [comment]
@maxwell once #6586 is merged, I'll adapt the cleaner PR #6471 to match remaining signatures. After that it's just a question of merging in that PR.

@cmrd-senya it still makes sense since removing some signatures only removes those columns. The remaining ones will still take a huge amount of space on pods like joindiaspora.com.

**Comment by cmrd-senya** [comment]
Well, actually, we can drop all data from the ```author_signature``` column when a relayable (comment, participation or like) is being stored on the pod different from the pod where parent entry is. That would reduce space usage significatly (and that would probably be enough). And we won't need any clearing after that, because all the signatures left would be required for the case of comments (relayables) fetch by a new pod as it has been said in https://github.com/diaspora/diaspora/issues/4920#issuecomment-163794165 by @SuperTux88.

**Comment by denschub** [comment]
This is a dupe of #4920. Please continue the discussion over there to avoid messing up conversations.

**Comment by jaywink** [comment]
@cmrd-senya sure, if we get far enough with changes to how data is saved and purge data that wasn't required to be saved in the first place (thanks for the excellent work re this btw) - then no point in running cleaning jobs. But, it's not merged yet, so I'll keep my PR open :)

**Comment by denschub** [comment]
Do you feel like I am making jokes or something? Please continue in the other ticket. Thanks.

**Comment by jaywink** [comment]
Dennis, why don't you check the timestamp of the comments. Or at least don't ever try to become a comedian.
